mdal/frmts/mdal_ply.cpp

```diff
@@ -0,0 +1,445 @@
+/*
+ MDAL - Mesh Data Abstraction Library (MIT License)
+ Copyright (C) 2020 Lutra Consulting Ltd.
+*/
+
+#include <stddef.h>
+#include <iosfwd>
+#include <iostream>
+#include <fstream>
+#include <sstream>
+#include <string>
+#include <memory>
+#include <vector>
+#include <map>
+#include <cassert>
+#include <limits>
+#include <algorithm>
+#include <string.h>
+
+#include "mdal_ply.hpp"
+#include "mdal.h"
+#include "mdal_utils.hpp"
+#include "mdal_logger.hpp"
+#include "mdal_data_model.hpp"
+#include "mdal_memory_data_model.hpp"
+
+#define DRIVER_NAME "PLY"
+
+MDAL::DriverPly::DriverPly() :
+  Driver( DRIVER_NAME,
+          "Stamford PLY Ascii Mesh File",
+          "*.ply",
+          Capability::ReadMesh
+        )
+{
+}
+
+MDAL::DriverPly *MDAL::DriverPly::create()
+{
+  return new DriverPly();
+}
+
+MDAL::DriverPly::~DriverPly() = default;
+
+size_t MDAL::DriverPly::getIndex( std::vector<std::string> v, std::string in )
+{
+  auto it = find( v.begin(), v.end(), in );
+  return ( size_t )distance( v.begin(), it );
+}
+
+// check for the magic number which in  a PLY file is "ply"
+bool MDAL::DriverPly::canReadMesh( const std::string &uri )
+{
+  std::ifstream in( uri, std::ifstream::in );
+  std::string line;
+  if ( !MDAL::getHeaderLine( in, line ) || !startsWith( line, "ply" ) )
+  {
+    return false;
+  }
+  return true;
+}
+
+std::unique_ptr<MDAL::Mesh> MDAL::DriverPly::load( const std::string &meshFile, const std::string & )
+{
+  MDAL::Log::resetLastStatus();
+
+  std::ifstream in( meshFile, std::ifstream::in );
+  std::string line;
+
+  // Read header
+  size_t vertexCount = 0;
+  size_t faceCount = 0;
+  size_t edgeCount = 0;
+  std::vector<std::string> chunks;
+  std::vector<MDAL::DriverPly::element> elements; // we will decant the element specifications into here
+  std::string proj = "";
+
+  if ( !std::getline( in, line ) )
+  {
+    MDAL::Log::error( MDAL_Status::Err_IncompatibleMesh, name(), meshFile + " Header is corrupt" );
+    return nullptr;
+  }
+
+  //
```

Comment:
please use /* */ style for multi-level comments

Reply:
Done in the next commit